python/runtime/feature_derivation/feature_derivation.py

```diff
@@ -0,0 +1,630 @@
+            field name.
+        selected_field_names (list[str]): the selected field name of the SQL
+            statement.
+        label_name (str): the label name of the TO TRAIN statement.
+
+    Returns:
+        None.
+    """
+    for target in targets:
+        if target not in fc_map:
+            fc_map[target] = {}
+
+        fc_target_map = fc_map[target]
+
+        new_fc_target_map = {}  # field_name -> list(FeatureColumn)
+        for field_name in fc_target_map:
+            if field_name in selected_field_names:
+                new_fc_target_map[field_name] = fc_target_map[field_name]
+                continue
+
+            if len(fc_map) > 1:
+                raise ValueError("cannot expand '%s' in COLUMN clause",
+                                 field_name)
+
+            field_pattern = re.compile(field_name, flags=re.I)
+            match_field_name = None
+            for selected_field_name in selected_field_names:
+                if field_pattern.fullmatch(selected_field_name):
+                    assert match_field_name is None, "%s matches duplicate fields" % field_name
+                    match_field_name = selected_field_name
+
+            if match_field_name is None:
+                raise ValueError(
+                    "'%s' in COLUMN clause does not match any selected fields"
+                    % field_name)
+
+            new_fc = fc_target_map[match_field_name][
+                0].new_feature_column_from(fd_map[match_field_name])
+            new_fc_target_map[match_field_name] = [new_fc]
+            del fd_map[field_name]
+
+        # ================== MAIN LOOP ==================
+        # Update or generate FeatureColumn for each selected field:
+        for selected_field_name in selected_field_names:
+            if label_name == selected_field_name:
+                continue  # ignore label field
+
+            fc_list = new_fc_target_map.get(selected_field_name)
+            if fc_list is not None:
+                for fc in fc_list:
+                    update_feature_column(fc, fd_map)
+            else:
+                if len(fc_map) > 1:
+                    # if column clause have more than one target, each target should specify the
+                    # full list of the columns to use.
+                    continue
+
+                field_desc = fd_map[selected_field_name]
+                if field_desc is None:
+                    raise ValueError("column not found or inferred: %s" %
+                                     selected_field_name)
+                new_fc = new_feature_column(field_desc)
+                new_fc_target_map[selected_field_name] = [new_fc]
+
+        fc_target_map.clear()
+        fc_target_map.update(new_fc_target_map)
+
+
+def transform_derived_feature_column_map_to_ir_feature_column_map(
+    features, fc_map, selected_field_names, label_name):
+    """
+    Transform the derived FeatureColumn map to the FeatureColumn
```

Comment:
Can we add more description about what transform is done in this method?
From the comments, we can see that the types of input and output are the same. It may be reader friendly if we have the description above.

Reply:
Added. Please see https://github.com/sql-machine-learning/sqlflow/pull/2744 .